openedx_learning/apps/authoring/publishing/admin.py

```diff
@@ -246,3 +250,261 @@ def get_queryset(self, request):
+    """
+    model = ContainerVersion
+    ordering = ["-publishable_entity_version__version_num"]
+    fields = [
+        "uuid",
+        "version_num",
+        "title",
+        "children",
+        "created",
+        "created_by",
+    ]
+    readonly_fields = fields  # type: ignore[assignment]
+    extra = 0
+
+    def get_queryset(self, request):
+        return super().get_queryset(request).select_related(
+            "publishable_entity_version"
+        )
+
+    def children(self, obj: ContainerVersion):
+        return _entity_list_detail_link(obj.entity_list)
+
+
+@admin.register(Container)
+class ContainerAdmin(ReadOnlyModelAdmin):
+    """
+    Django admin configuration for Container
+    """
+    list_display = ("key", "uuid", "created", "draft", "published", "see_also")
+    fields = [
+        "publishable_entity",
+        "learning_package",
+        "draft",
+        "published",
+        "created",
+        "created_by",
+        "see_also",
+        "most_recent_parent_entity_list",
+    ]
+    readonly_fields = fields  # type: ignore[assignment]
+    search_fields = ["publishable_entity__uuid", "publishable_entity__key"]
+    inlines = [ContainerVersionInlineForContainer]
+
+    def uuid(self, obj: Container) -> SafeText:
+        return model_detail_link(obj, obj.uuid)
+
+    def learning_package(self, obj: Container) -> SafeText:
+        return model_detail_link(
+            obj.publishable_entity.learning_package,
+            obj.publishable_entity.learning_package.key,
+        )
+
+    def get_queryset(self, request):
+        return super().get_queryset(request).select_related(
+            "publishable_entity",
+            "publishable_entity__learning_package",
+            "publishable_entity__published__version",
+            "publishable_entity__draft__version",
+        )
+
+    def draft(self, obj: Container) -> SafeText:
+        """
+        Link to this Container's draft ContainerVersion
+        """
+        if draft := obj.versioning.draft:
+            return format_html(
+                "Version {} ({})", draft.version_num, _entity_list_detail_link(draft.entity_list)
+            )
+        return SafeText("-")
+
+    def published(self, obj: Container) -> SafeText:
+        """
+        Link to this Container's published ContainerVersion
+        """
+        if published := obj.versioning.published:
+            return format_html(
+                "Version {} ({})", published.version_num, _entity_list_detail_link(published.entity_list)
+            )
+        return SafeText("-")
+
+    def see_also(self, obj: Container):
+        return one_to_one_related_model_html(obj)
+
+    def most_recent_parent_entity_list(self, obj: Container) -> SafeText:
+        if latest_row := EntityListRow.objects.filter(entity_id=obj.publishable_entity_id).order_by("-pk").first():
+            return _entity_list_detail_link(latest_row.entity_list)
+        return SafeText("-")
```

Comment:
Is there any reason to wrap these hyphens in `SafeText` ?

Reply:
It was for type consistency. But, it looks like mypy understands that SafeText is a subtype of str, so I can just change the return type to str and mypy is happy with me just returning `"-"`.